### PR TITLE
Trip Status UI

### PR DESCRIPTION
**Pull request by evansiroky**
This PR depends on https://github.com/ibi-group/otp-middleware/pull/124.

This PR adds a trip status UI for monitored trips in the edit trip view. It also moves all monitored trip files into the `lib/components/user/monitored-trip` folder.

In accordance with discussions internally at IBI Group, the status UI is shown as a part of the edit trip view. The Trip Status is always shown with 4 sections: Header, Body, Alerts and Footer Buttons.

Screen Shot 2020-12-14 at 2 50 14 AM

### Overall Panel

The overall structure of the trip status UI is always the same. The text and buttons differ based on a combination of the trip status, whether snoozing or pausing exists, and realtime info if the trip is being actively monitored. The overall status view is contained in a Bootstrap Panel and has a Bootstrap style set dynamically.

### Header

The header always has some kind of header text and then a footnote about when the last update occurred.

### Body

The body of the panel will always have some kind of text describing more details about the trip.

### Alerts

This isn't always shown. It is only shown when a trip is being actively monitored and at least one alert exists in the matching itinerary. The display of the alerts use a component from the `@opentripplanner/itinerary-body` package.

### Footer Buttons

Four buttons (usually just 2 at once) are conditionally shown based on various states of the monitored trip. They can be:

- Snooze/Unsnooze Trip
- Pause/Resume Trip
- Delete Trip
- Plan New Trip

### CODE REVIEW

File: lib/components/user/monitored-trip/trip-status.js
Comment (landonreed) [inline lib/components/user/monitored-trip/trip-status.js:391]: This file is pretty mondo. I don't see any obvious ways to split it up outside of splitting up the header, body, alerts, and footer buttons, but I think that might be warranted.

**Review by landonreed — CHANGES_REQUESTED** [review]
Just a few minor comments to start with, but I still need to explore this a bit more.

**Comment by binh-dam-ibigroup** [comment]
This PR also depends on #291.

File: lib/components/user/monitored-trip/trip-status.js
Comment (binh-dam-ibigroup) [inline lib/components/user/monitored-trip/trip-status.js:122]: The conditions the block below are similar to the if block above, but I don't see a good way to refactor.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Consider #295 for refactoring some of the rendering logic, and see other comments as indicated.

File: lib/components/user/monitored-trip/trip-status.js
Comment (binh-dam-ibigroup) [reply lib/components/user/monitored-trip/trip-status.js:112]: How about this color scheme @landonreed  and @evansiroky:
- Grey for upcoming trips until the lead time until departure
- Blue for trips without realtime updates past the lead time until departure (including active trips)
- Green for trips with realtime updates past the lead time until departure (including active trips)
- Other colors remain as is.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: I've added the `notYetCalculatedTripRenderer` which should address this.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: This is now outdated after refactoring.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: I also didn't quite like how spaghetti this file was and spent a while thinking about how to best organize it both before you wrote this comment and now again after. I have gone ahead and significantly refactored this component into a hopefully more comprehensible structure.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: I'm going to hold off on any further updates since that might require custom styling. This code currently just sets a bootstrap panel theme and the rest gets taken care of.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: Since these are standard HTML tags encapsulated within a bootstrap panel, they don't lack styling rules completely, so there is some styling that gets applied. Not sure what other custom styles we need, so I'm not going to implement anything around this for now.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: These have been removed after a refactor.

File: lib/actions/user.js
Comment (evansiroky) [reply lib/actions/user.js:398]: I believe I tried @binh-dam-ibigroup's idea and it still didn't work.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: Added in 7d5505703026f41c849998ada8d6a66278f321b5

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: Added in 7d5505703026f41c849998ada8d6a66278f321b5

File: lib/actions/user.js
Comment (evansiroky) [reply lib/actions/user.js]: Renamed in 7d5505703026f41c849998ada8d6a66278f321b5

File: lib/components/user/monitored-trip/saved-trip-screen.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/saved-trip-screen.js]: Alphabetized in 7d5505703026f41c849998ada8d6a66278f321b5

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: Added in 7d5505703026f41c849998ada8d6a66278f321b5

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: I believe I have updated to this color scheme in 7d5505703026f41c849998ada8d6a66278f321b5

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: No, because since they are in upcoming trips, it is possible that these alerts are now stale due to the trip not having been calculated in a while. Therefore, they shouldn't be shown until after the lead time. I added a comment about this in 7d5505703026f41c849998ada8d6a66278f321b5

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: Removed in 7d5505703026f41c849998ada8d6a66278f321b5

**Comment by evansiroky** [comment]
As noted in some comments, I have significantly refactored this PR so that it utilizes the strategy pattern to generate most of the necessary dynamic inputs into the trip status component. So, this is ready for review again.

File: lib/components/user/monitored-trip/trip-basics-pane.js
Comment (binh-dam-ibigroup) [inline lib/components/user/monitored-trip/trip-basics-pane.js:106]: Do we want the trip status to be displayed when the user is saving the itinerary for the first time? (At that point, the trip has not been created in the database.)

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Before I further review the changes, I need to point out the confusion between the monitoring start time and the trip start time.

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/index.js
Comment (landonreed) [inline lib/components/user/monitored-trip/trip-status-rendering-strategies/index.js:41]: Do we want to throw an error like this and risk a white screen of death (we don't have any error boundaries built in to the app right now)? I think a better approach would be to just fill those fields with some values (e.g., "Unknown status" or something).

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/upcoming-trip-renderer.js
Comment (landonreed) [inline lib/components/user/monitored-trip/trip-status-rendering-strategies/upcoming-trip-renderer.js:83]: This logic could be confusing to someone. Might be good to reference the assignment to `shouldRenderAlerts` above in a comment here.

**Review by landonreed — CHANGES_REQUESTED** [review]
This refactor looks really good, Evan. Nice work. Just a couple of minor comments. I haven't spun this up and played around with all of the states because I'm not sure of the best way to trigger everything (follow on from our Tuesday conversation), but I have these comments for now.

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js
Comment (binh-dam-ibigroup) [inline lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js:4]: This is a good concept. The term `renderer` doesn't quite fit because the function doesn't actually return any JSX, but I can't think of an alternate right now...

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js
Comment (binh-dam-ibigroup) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js:9]: That's why suggested adding a TODO somewhere else for this PR to sync the thresholds. Right now we use 60 seconds in `ItineraryBody` and `StopViewer`.

File: lib/components/user/monitored-trip/trip-status.js
Comment (binh-dam-ibigroup) [inline lib/components/user/monitored-trip/trip-status.js:72]: Sorry I'm doubling down on this since it is not a heading... `<div>` and/or `<small>`?

**Review by binh-dam-ibigroup — COMMENTED** [review]
See additional minor comments, in addition to my change request earlier.

File: lib/components/user/monitored-trip/trip-status.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status.js]: Changed to `<small>` in 8f071e25ea62586b9d67fbef14430316d2ae2abe

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/upcoming-trip-renderer.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/upcoming-trip-renderer.js]: Since the default is to show alerts, I was able to remove this entirely in 8f071e25ea62586b9d67fbef14430316d2ae2abe

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/index.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/index.js]: I added these as defaults in the base renderer in 8f071e25ea62586b9d67fbef14430316d2ae2abe

File: lib/actions/user.js
Comment (evansiroky) [reply lib/actions/user.js]: Renamed in 8f071e25ea62586b9d67fbef14430316d2ae2abe

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js:8]: JSDoc added in 9df85f9f795ca1cc1964576f212bf0a6076246b6

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js
Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js:13]: I've gone ahead and made another PR to address this. See https://github.com/opentripplanner/otp-react-redux/pull/305. Thanks, @binh-dam-ibigroup for the tip on where to find the other components that handle delays.

**Comment by evansiroky** [comment]
Ready for review again.

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/upcoming-trip-renderer.js
Comment (binh-dam-ibigroup) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/upcoming-trip-renderer.js:37]: @evansiroky The highlighted lines above result in extra space as shown below:
image

**Review by binh-dam-ibigroup — APPROVED** [review]
Conditional approval, please address these two comments from a previous review before merging:
- https://github.com/opentripplanner/otp-react-redux/pull/290/files#r558444095
- https://github.com/opentripplanner/otp-react-redux/pull/290/files#r556617329

Thanks for the changes!

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js
Comment (binh-dam-ibigroup) [inline lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js:17]: Typos 'Unknown'... The spellchecker didn't catch these?

Comment (evansiroky) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/base-renderer.js]: That is unfortunate that the spellchecker didn't catch these. I'll add it to my todo-list to explore other spellcheckers. For now, I fixed this in 3a36d3983359b49803f6cd8df4a6d2eb49c83251

**Review by landonreed — APPROVED** [review]
Really great stuff, @evansiroky.

**Pull request merged**
